codebuild/bin/grep_simple_mistakes.sh

```diff
@@ -14,9 +14,48 @@
 
 FAILED=0
 
+#############################################
+# Grep for any instances of raw memcmp() function. s2n code should instead be
+# using s2n_constant_time_equals()
+#
+# KNOWN_MEMCMP_USAGE is used to capture all known uses of memcmp and acts as a
+# safeguard against any new uses of memcmp.
+#############################################
+S2N_FILES_ASSERT_NOT_USING_MEMCMP=$(find "$PWD" -type f -name "s2n*.[ch]" -not -path "*/tests/*")
+declare -A KNOWN_MEMCMP_USAGE
+KNOWN_MEMCMP_USAGE["$PWD/crypto/s2n_rsa.c"]=1
+KNOWN_MEMCMP_USAGE["$PWD/tls/s2n_early_data.c"]=1
+KNOWN_MEMCMP_USAGE["$PWD/tls/s2n_kem.c"]=1
+KNOWN_MEMCMP_USAGE["$PWD/tls/s2n_cipher_suites.c"]=3
+KNOWN_MEMCMP_USAGE["$PWD/tls/s2n_server_hello.c"]=4
+KNOWN_MEMCMP_USAGE["$PWD/tls/s2n_security_policies.c"]=1
+KNOWN_MEMCMP_USAGE["$PWD/tls/s2n_psk.c"]=1
+KNOWN_MEMCMP_USAGE["$PWD/tls/s2n_config.c"]=1
+KNOWN_MEMCMP_USAGE["$PWD/tls/s2n_resume.c"]=2
+KNOWN_MEMCMP_USAGE["$PWD/tls/s2n_connection.c"]=1
+KNOWN_MEMCMP_USAGE["$PWD/tls/s2n_protocol_preferences.c"]=1
+KNOWN_MEMCMP_USAGE["$PWD/utils/s2n_map.c"]=3
+KNOWN_MEMCMP_USAGE["$PWD/stuffer/s2n_stuffer_text.c"]=1
+
+for file in $S2N_FILES_ASSERT_NOT_USING_MEMCMP; do
+  # NOTE: this matches on 'memcmp', which will also match comments. However, there
+  # are no uses of 'memcmp' in comments so we opt for this stricter check.
+  RESULT_NUM_LINES=`grep -n 'memcmp' $file | wc -l`
+  # check if memcmp usage is 0 or a known value
+  if [ "${RESULT_NUM_LINES}" != "0" ] && [ "${RESULT_NUM_LINES}" != "${KNOWN_MEMCMP_USAGE["$file"]}" ]; then
```

Comment:
yep good catch.. ill add that check

Reply:
total memcmp check:
failure: https://github.com/aws/s2n-tls/pull/3059/commits/1f8f2a12b32d5375b22cd7a4331ea5426074137c
success: https://github.com/aws/s2n-tls/pull/3059/commits/d4834fc88bdbbb3d0ee8d26c93bc139183bf9270

Reply:
That's going to make debugging difficult / fixing the error difficult, since it doesn't tell you what file is wrong. Why not just compare every file's results to either the entry in the map (if it exists) or 0? Remove the `"${RESULT_NUM_LINES}" != "0"` requirement.

Reply:
simplified the logic as you suggested.

we now check against the map for each file and confirm that the value matches or is 0